doc/designs/quic-design/quic-io-arch.md

```diff
@@ -0,0 +1,485 @@
+  way to force these calls to return once `SSL_free` is called and we need to
+  tear down the connection. However, the BIO interface does not provide
+  any way to do this. *At best* we might assume the BIO is a `BIO_s_dgram`
+  (but cannot assume this in the general case), but even then we can only
+  accomplish teardown by violating the BIO abstraction and closing the
+  underlying socket.
+
+  This is the only portable way to ensure that a recv(3) call to the same socket
+  returns. This obviously is a highly application-visible change (and is likely
+  to be far more disruptive than configuring the socket into non-blocking mode).
+
+  Moreover, it is not workable anyway because it only works for a socket-based
+  BIO and violates the BIO abstraction. For BIOs in general, there does not
+  appear to be any viable solution to the teardown issue.
+
+Even if this approach were successfully implemented, applications will still
+need to change to using network BIOs with datagram semantics. For applications
+using custom BIOs, this is likely to require substantial rework of those BIOs.
+There is no possible way around this. Thus, even if this solution were adopted
+(notwithstanding the issues which preclude this noted above) for the purposes of
+accommodating applications using custom network BIOs in a blocking mode, these
+applications would still have to completely rework their implementation of those
+BIOs. In any case, it is expected to be very rare that sophisticated
+applications implementing their own custom BIOs will do so in a blocking mode.
+
+### Use of non-blocking I/O
+
+By comparison, use of non-blocking I/O and select(3) or similar APIs on the
+network side makes satisfying our requirements for QUIC easy, and also allows
+our internal approach to I/O to be flexibly adapted in the future as
+requirements may evolve.
+
+This is also the approach used by all other known QUIC implementations; it is
+highly unlikely that any QUIC implementations exist which use blocking network
+I/O, as (as mentioned above) it would lead to suboptimal performance due to the
+ACK delay issue.
+
+Note that this is orthogonal to whether we provide blocking I/O semantics to the
+application. We can use blocking I/O internally while using this to provide
+either blocking or non-blocking semantics to the application, based on what the
+application requests.
+
+This approach in general requires that a network socket be configured in
+non-blocking mode. Though some OSes support a `MSG_DONTWAIT` flag which allows a
+single I/O operation to be made non-blocking, not all OSes support this (e.g.
+Windows), thus this cannot be relied on. As such, we need to configure any
+socket FD we use into non-blocking mode.
+
+Of the approaches outlined in this document, the use of non-blocking I/O has the
+fewest disadvantages and is the only approach which appears to actually be
+implementable in practice. Moreover, each disadvantage can be readily mitigated:
+
```

Comment:
A disadvantage is the additional complexity of needing the poll descriptor interface.

Reply:
Updated to note this.